### PR TITLE
Use `io.flutter.Build.API_LEVELS` rather than `android.os.Build.VERSION_CODES`

### PR DESCRIPTION
**Pull request by dnfield**
Updates the linting script to ban the use of `VERSION_CODES`.

We currently have a mish-mash of using the integers, using `VERSION_CODES`, and even how we import the version codes. This makes it more confusing when doing things like #51070 - I think it is clearer to see `22` than `LOLLIPOP_MR1`.

I'd like to get LGTM (or at least no opinion) from all the requested reviewers here.

### CODE REVIEW

**Review by matanlurey — COMMENTED** [review]
As long as we're consistent one way or another (and it works with our linting strategy), I'm LGTM.

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java [hunk 1]
```diff
@@ -758,7 +758,7 @@ public AccessibilityNodeInfo createAccessibilityNodeInfo(int virtualViewId) {
       result.setParent(rootAccessibilityView, semanticsNode.parent.id);
     } else {
       if (BuildConfig.DEBUG && semanticsNode.id != ROOT_NODE_ID) {
-        Log.e(TAG, "Semantics node id does not equal ROOT_NODE_ID.");
+        Log.e(TAG, "Sema_ntics node id does not equal ROOT_NODE_ID.");
```
Comment (gmackall) [inline shell/platform/android/io/flutter/view/AccessibilityBridge.java:761]: Should this change be reverted?

**Review by gmackall — COMMENTED** [review]
I'm fine either way (but agreed we should have only one way). I find the integers more readable but I also like that the codes are more easily searchable

**Review by camsim99 — COMMENTED** [review]
Also LGTM! Glad to have something consistent moving forward.

File: shell/platform/android/io/flutter/view/AccessibilityBridge.java (hunk 1 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/view/AccessibilityBridge.java:761]: Yes, thank you

**Comment by johnmccutchan** [comment]
This seems like a step backwards in terms of code readability.

Before: "Build.VERSION_FOO"
After: "18"

We lose our ability to search for something meaningful.

Why are we making this change?

At the very least, why don't we introduce our own "AndroidBuild.VERSION_N" for all N and use that instead of the raw integer?

**Review by johnmccutchan — CHANGES_REQUESTED** [review]
See my comment on the issue.

**Comment by dnfield** [comment]
> At the very least, why don't we introduce our own "AndroidBuild.VERSION_N" for all N and use that instead of the raw integer?

I think this works for me - but I'd want the raw integer in there, e.g. `io.flutter.Build.VERSION_CODES.API_24`, _not_ whatever the Android team decided to use as a maybe single letter or maybe whole word or maybe service release code in a given year.

**Comment by dnfield** [comment]
I have two reasons for wanting this change:

1. Consistency. Right now we use a mixture of integers and `VERSION_CODES`. We also use different style for how we import the version codes.
2. The numbers correspond to what the Android docs actually use - e.g. they will say `Introduced in API 28`, not `Introduced in Android P`.
3. It will be harder to lint that we do not use integers and keep introducing the inconsistency - although we could at least lint that you use our special class and not the `VERSION_CODES` class from Android.

**Review by gmackall — APPROVED** [review]
LGTM! Glad we found a solution that maximizes both searchability and readability

**Comment by johnmccutchan** [comment]
Thanks for making my requested changes @dnfield

File: shell/platform/android/io/flutter/Build.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter;
+
+/** A replacement of utilities from android.os.Build. */
+public class Build {
+  /** For use in place of the Android VERSION CODES class. */
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/Build.java:9]: Rewrite the comment so if someone searched for the official class they would find this:

/** For use in place of Build.VERSION_CODES class */

File: shell/platform/android/io/flutter/plugin/mouse/MouseCursorPlugin.java [hunk 3]
```diff
@@ -5,16 +5,15 @@
...
 import android.view.PointerIcon;
 import androidx.annotation.NonNull;
 import androidx.annotation.RequiresApi;
 import io.flutter.embedding.engine.systemchannels.MouseCursorChannel;
 import java.util.HashMap;
 
 /** A mandatory plugin that handles mouse cursor requests. */
-@TargetApi(Build.VERSION_CODES.N)
-@RequiresApi(Build.VERSION_CODES.N)
+@TargetApi(24)
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/plugin/mouse/MouseCursorPlugin.java:15]: Fix here and below

File: shell/platform/android/io/flutter/plugin/text/ProcessTextPlugin.java [hunk 4]
```diff
@@ -134,8 +136,8 @@ private void cacheResolveInfos() {
    * <p>When an activity does not return a value. the request is completed successfully and returns
    * null.
    */
-  @TargetApi(Build.VERSION_CODES.M)
-  @RequiresApi(Build.VERSION_CODES.M)
+  @TargetApi(23)
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/plugin/text/ProcessTextPlugin.java:139]: fix here and below

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 5]
```diff
@@ -789,8 +791,8 @@ private void releaseAccessibilityNodeProvider() {
   // -------- Start: Mouse -------
 
   @Override
-  @TargetApi(Build.VERSION_CODES.N)
-  @RequiresApi(Build.VERSION_CODES.N)
+  @TargetApi(24)
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/view/FlutterView.java:794]: fix here and here

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java [hunk 6]
```diff
@@ -253,7 +255,7 @@ private void closeCurrentImage() {
 
   @TargetApi(29)
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:256]: fix

File: shell/platform/android/test/io/flutter/plugin/mouse/MouseCursorPluginTest.java [hunk 7]
```diff
@@ -25,7 +26,7 @@
 
 @Config(
     manifest = Config.NONE,
-    minSdk = 24,
+    minSdk = API_LEVELS.API_24,
     shadows = {})
 @RunWith(AndroidJUnit4.class)
 @TargetApi(24)
```
Comment (johnmccutchan) [inline shell/platform/android/test/io/flutter/plugin/mouse/MouseCursorPluginTest.java:32]: fix

File: shell/platform/android/test/io/flutter/plugin/platform/SingleViewPresentationTest.java [hunk 8]
```diff
@@ -24,10 +23,10 @@
 
 @Config(manifest = Config.NONE)
 @RunWith(AndroidJUnit4.class)
-@TargetApi(P)
+@TargetApi(28)
```
Comment (johnmccutchan) [inline shell/platform/android/test/io/flutter/plugin/platform/SingleViewPresentationTest.java:26]: fix

File: shell/platform/android/test/io/flutter/plugin/text/ProcessTextPluginTest.java [hunk 9]
```diff
@@ -36,8 +35,8 @@
 import org.mockito.ArgumentCaptor;
 
 @RunWith(AndroidJUnit4.class)
-@TargetApi(Build.VERSION_CODES.N)
-@RequiresApi(Build.VERSION_CODES.N)
+@TargetApi(24)
```
Comment (johnmccutchan) [inline shell/platform/android/test/io/flutter/plugin/text/ProcessTextPluginTest.java:38]: fix here and below

File: shell/platform/android/io/flutter/Build.java (hunk 2 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/Build.java:9]: I can do that, but I'll have to special case the forbidden imports script to allow it in this file.

**Review by johnmccutchan — APPROVED** [review]

File: shell/platform/android/io/flutter/Build.java (hunk 2 above)
Comment (johnmccutchan) [reply shell/platform/android/io/flutter/Build.java]: Can the linter detect that text is contained inside of a comment?

File: testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/ExternalTextureTests.java [hunk 10]
```diff
@@ -82,7 +81,7 @@ public void testRotatedMediaSurface_270() throws Exception {
   }
 
   @Test
-  @SdkSuppress(minSdkVersion = VERSION_CODES.M)
+  @SdkSuppress(minSdkVersion = 23)
```
[the remark this answers is not in the document]
Comment (dnfield) [reply testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/ExternalTextureTests.java:84]: Fixed.

File: shell/platform/android/test/io/flutter/view/AccessibilityBridgeTest.java [hunk 11]
```diff
@@ -81,7 +82,7 @@ public void itDescribesNonTextFieldsWithAContentDescription() {
   }
 
   @Config(sdk = 28)
```
[the remark this answers is not in the document]
Comment (dnfield) [reply shell/platform/android/test/io/flutter/view/AccessibilityBridgeTest.java:84]: Fixed.

File: tools/android_illegal_imports.py [hunk 12]
```diff
@@ -42,22 +44,29 @@ def main():
...
-      bad_trace_files, 'android[x].tracing.Trace', FLUTTER_TRACE_CLASS
-  )
+      if ANDROID_BUILD_VERSION_CODE_CLASS in contents:
+        bad_version_codes_files.append(file)
+
+  has_bad_files = CheckBadFiles(bad_log_files, ANDROID_LOG_CLASS,
+                                FLUTTER_LOG_CLASS) or CheckBadFiles(
+                                    bad_trace_files, 'android[x].tracing.Trace', FLUTTER_TRACE_CLASS
+                                ) or CheckBadFiles(
+                                    bad_version_codes_files, 'android.os.Build.VERSION_CODES',
```
[the remark this answers is not in the document]
Comment (dnfield) [reply tools/android_illegal_imports.py:74]: Added some comments.

**Pull request merged**